File: Java/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        
        if (head == null){
            return false;
        }
        
        ListNode slowPtr = head;
        ListNode fastPtr = head;
        
        /** Approach
        
        // 1. if no cycle exists AND
            // a. if there are even number of nodes then fast pointer will be pointing to null.
            // b. else there are odd number of nodes then fast pointer will be pointing to last node.
        
        // 2. if slow pointer meet fast pointer that interprets there exists a loop 
        */
        
        while (fastPtr != null && fastPtr.next != null) {
            
            // move slow pointer one node at a time
            slowPtr = slowPtr.next;
            
            // move fast pointer two nodes at a time
            fastPtr = fastPtr.next.next;
            
            if (slowPtr == fastPtr)
                return true;
        }
        
        return false;
    }
}

File: Python/Linked_List_Cycle.py
#Time: 95.48%
#Space: 99.98%
#Difficulty: Easy
#Acceptance: 49.0%

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Return False if the list is empty or has only one node (no cycle possible)
        if not head or not head.next:
            return False
        
        # Initialize two pointers: 'slow' moves one step at a time, 'fast' moves two steps
        slow = head
        fast = head.next

        # Iterate through the list
        while slow != fast:
            # If 'fast' reaches the end of the list (or its next step is None), there's no cycle
            if not fast or not fast.next:
                return False

            # Move 'slow' one step forward and 'fast' two steps forward
            slow = slow.next
            fast = fast.next.next

        # If 'slow' and 'fast' meet, there is a cycle in the list
        return True
